### DISCUSSION TITLE
TXT record is not found by LE

### DISCUSSION THREAD

**Discussion by TatyanaBol** [go-acme/lego#1798] [Q&A]
We provide a service that orders certificates from LE.
We use Lego library,  DNS-01 challenges.
A customer creates a config with his DNS provider credentials and we make everything for him.
Our code implements functions _Present_, _CleanUp_, it adds/removes the TXT record through DNS provider API
Usually everything works, we have a lot of customers.

Recently orders started failing for **one specific customer**, he uses Softlayer DNS provider.
In our log, I can see that a challenge was set, but when Lego calls LE it returns the error `"acme: error: 400 :: urn:ietf:params:acme:error:dns :: DNS problem: NXDOMAIN looking up TXT for _acme-challenge.logs-epfp01-00465.qradar.ibmcloud.com - check that a DNS record exists for this domain"`.   

We increased the polling interval to wait between the set challenge and validation to 10 minutes for case of slow propagation. The customer complains that 7 of 15 orders still fail.

Please help to understand what is the issue.
Does Lego poll both authoritative DNS servers?
What can be a reason that Lego finds TXT record and LE not?

Thanks in advance

**Comment by ldez** [comment]
Hello,

Lego and LE don't use the same DNS servers because LE doesn't communicate publically on the DNS servers they use.

The DNS propagation can be very very long with some DNS providers, but we check the propagation of the TXT record and if the domain doesn't exist the check will stop before the end of the propagation check.

The `NXDOMAIN` error seems to indicate that the A record is very slow to propagate.
So I recommend your customer checks the propagation of the A records before trying to get a certificate.
 
A bit off-topic but I recommend using CNAME instead of asking for the DNS credentials of your customers.
https://letsencrypt.org/2019/10/09/onboarding-your-customers-with-lets-encrypt-and-acme.html#the-advantages-of-a-cname

**Reply by TatyanaBol** [reply]
> The NXDOMAIN error seems to indicate that the A record is very slow to propagate.
> So I recommend your customer checks the propagation of the A records before trying to get a certificate.

Why are you talking about A records when we set TXT records?
Lego checks the propagation and get a certificate for us, should we add our code to check propagation before Lego checks it?

**Reply by ldez** [reply]
because NXDOMAIN means Non eXistent Domain.

So either the domain doesn't exist or the A (or AAAA) record is missing.

**Reply by TatyanaBol** [reply]
Lets say a customer has mydomain.com
He orders a certificate for aaa.bbb.mydomain.com
We add  the TXT record `_acme-challenge.aaa.bbb.domain.com` in mydomain.com

Which A (or AAAA)  record is missing?
Who and when should create this record?
Why a challeng is OK for Lego, doesn't Lego check this A record?

**Reply by ldez** [reply]
It's the base of domain management: an A record is a record that points to the IP of the targeted server/application.

The A record is not related to lego but to a domain, to be able to use it as a domain.
Every public domain should have an A (ipv4) and/or AAAA (ipv6) record.

I was thinking that you understand what NXDOMAIN domain means.

`_acme-challenge.aaa.bbb.domain.com` doesn't need an A record but the domain/zone needs one.

NXDOMAIN is the answer of a DNS server to say that a domain doesn't exist.

**Reply by ldez** [reply]
If you need customer support, as you have a lot of customers, don't hesitate to ask to IBM to sponsor me.

https://github.com/sponsors/ldez

**Reply by TatyanaBol** [reply]
If you mean that mydomain.com is not configured well then I should say that today the customer made 11 orders, 9 of them succeeded and 2 failed with the error above. 
The orders differed by subdomains:  for aaa.mydomain.com, bbb.mydomain.com, etc
Again, 9 of them succeeded and 2 failed

**Reply by ldez** [reply]
Sorry but it's not a problem with lego, it's a DNS problem: the sub-domains (`aaa.mydomain.com`, `bbb.mydomain.com`) don't exist in the DNS servers that are called.

**Reply by TatyanaBol** [reply]
`NXDOMAIN looking up TXT for _acme-challenge.domain - check that a DNS record exists for this domain"` means that TXT record is not found. Why are you talking about A record?
TXT record was found by Lego but not by LE for 2 orders of 9.

**Reply by ldez** [reply]
> NXDOMAIN looking up TXT for _acme-challenge.domain - check that a DNS record exists for this domain" means that TXT record is not found. 

It means that the domain (example.com, or _acme-challenge.example.com) doesn't exist.

The error means: when lego tries to find the TXT record, the DNS server answers that the domain doesn't exist.

**Reply by ldez** [reply]
In all cases, it's a DNS problem, for me, it's a propagation problem: either the DNS provider is slow to really create the records (A, AAAA, TXT), or the records (A, AAAA, TXT) are not propagated, or the records (A, AAAA, TXT) are slow to be propagated.

**Reply by TatyanaBol** [reply]
Sorry, maybe I didn't explain the problem well.
**Lego DOES FIND the TXT record from the first attempt**
LE after that doesn't find.
My question is - is there some way to improve Lego's propagation check to make a successful LE check more likely?

**Reply by ldez** [reply]
> My question is - is there some way to improve Lego's propagation check to make a successful LE check more likely?

Yes, you can play with the resolvers:

```
   --dns.resolvers value        Set the resolvers to use for performing recursive DNS queries. Supported: host:port. The default is to use the system resolvers, or Google's DNS resolvers if the system's cannot be determined.
```

https://go-acme.github.io/lego/usage/cli/options/

You can provide a set of DNS servers that lego will use to check the propagation.
But as I already said, the list of the DNS servers used by LE is not public.

FYI, I said "lego" just as an example.

**Reply by TatyanaBol** [reply]
Thank you!

**Reply by ldez** [reply]
give me some feedback if the `--dns.resolvers` fix your issue.

**Reply by TatyanaBol** [reply]
Does Lego **always** check a domain authoritative nameservers to have a challenge TXT record?
How would you recommend selecting `dns resolvers`?

**Reply by TatyanaBol** [reply]
@ldez 
If I understand correctly, Lego now checks a challenge, and when a TXT record is found, Lego immediately asks LE to validate the challenge. 
Can we add a configurable delay after the TXT record is found and before calling LE? Configurable like **polling interval** and **timeout**.
For "very slow" DNS providers like Softlayer, we would want to increase the delay, and for "fast" providers to configure it to be very small.

**Reply by ldez** [reply]
Lego always checks the propagation: with `dns-resolvers` you just choose the DNS servers instead of using the default ones.
There is no configurable delay between the check of the propagation and the call to LE because the check of the propagation is here to handle that problem.

**Reply by TatyanaBol** [reply]
We see that if Lego finds the TXT record, it doesn't mean that LE immediately sees it. We would want to wait for the proper propagation a little.
We wrote our js acme-client, and that's what we did there - the delay between our check and LE validation. Softlayer orders worked only with such delay.
Is it possible to add this parameter? With value 0 by default.

**Reply by ldez** [reply]
> We see that if Lego finds the TXT record, it doesn't mean that LE immediately sees it. 

The goal of the check of the propagation is to check the availability of the TXT record before calling LE. So if Lego finds the TXT record, LE should see it.

> Is it possible to add this parameter? With value 0 by default.

Your DNS provider must assume the right propagation, IBM Cloud (SoftLayer) must improve the propagation quality.

A delay will never be able to warranty the availability, so the option will be weak.

Adding a "weak workaround option" only for one provider doesn't sound good.

**Reply by ldez** [reply]
The only workaround I see is to use a large propagation interval and use `dns-servers` pointing to "slow to update" DNS servers.

**Reply by ldez** [reply]
Are you using lego as a library? (not as a binary)

**Reply by TatyanaBol** [reply]
Yes, as a library.
I see. We tried to increase propagation interval to 10 and even 12 minues, some part of orders still failed and others were very long.
Can you help to understand how we can know which DNS servers to use for dns-resolvers? How to choose?

**Reply by TatyanaBol** [reply]
it feels wrong that we increase `interval` and not one-time `delay`. 
So the polling works with intervals of 10-12 minutes, the orders are very long, and it still doesn't ensure that the orders will succeed.
We could check with interval 1-2 minutes, then wait 5 min and chances for success would be much higher.

**Reply by ldez** [reply]
> So the polling works with intervals of 10-12 minutes, the orders are very long, and it still doesn't ensure that the orders will succeed.

the delay will also not ensure that the orders will succeed, because it's a delay and not accurate information about the propagation.

You can add a custom `preCheck` via `dns01.WrapPreCheck()`: `preCheck` is the function called during the propagation check, so you can add an extra `sleep` when the check of the propagation succeeds.

**Reply by TatyanaBol** [reply]
Thanks! I'll look at it

**Reply by ldez** [reply]
Something like that:

```go
err = client.Challenge.SetDNS01Provider(provider,
	dns01.CondOption(len(resolvers) > 0, dns01.AddRecursiveNameservers(resolvers)),
	dns01.WrapPreCheck(func(domain, fqdn, value string, check dns01.PreCheckFunc) (bool, error) {
		if delayBeforeCheck > 0 {
			logger.Debugf("Delaying %d rather than validating DNS propagation now.", delayBeforeCheck)
			time.Sleep(time.Duration(delayBeforeCheck))
		}

		b, err := check(fqdn, value)
		if err == nil {
			logger.Debugf("Delaying %d ", extraDelay)
			time.Sleep(extraDelay)
		}

		return b, err
	}),
)
```

**Reply by TatyanaBol** [reply]
Thank you, we are going with your last advice.
